### DISCUSSION TITLE
Existence of createConsBasicLinear function

### DISCUSSION THREAD

**Discussion by Engberg-Jacob** [scipopt/PySCIPOpt#910] [Q&A] [answered]
Hello!

I'm examining another repo which uses pySCIPopt, and in this code they use a model attribute, namely `Model().createConsBasicLinear()`, and I can't get it to work, I can't find it in the pySCIPopt source code, neither the current version or earlier versions. So I just wanted to check whether I've completely dropped the ball on something here, or if this wrapper haven't been implemented yet.

(The reason I ask is that, in the SCIP documentation, `SCIPcreateConsBasicLinear()` does not require a custom constrainthandler, whilst `SCIPcreateCons()` does, so I don't see an easy way of one replacing the other)

Thank you!

**Comment by Joao-Dionisio** [comment]
Hey @Engberg-Jacob! Yeah, I don't think this was ever wrapped to PySCIPOpt. I need to discuss with the others to see whether it makes sense to interface. Can you please share what you're trying to do? Maybe there's a better way to do it with what we have in PySCIPOpt at the moment.

Also, can you share the repo you mentioned? I'd like to take a look in case there are other useful contributions :)

**Reply by Engberg-Jacob** [reply]
Sure! The repo is https://github.com/pandat8/ML4LB/tree/mpc - though I don't think it is actively being maintained. I am mostly just trying to run the code in the repo to reproduce the results, and take it from there, but run into issues. The code is modifying how to branch, and when trying to it specifically happens on line 1017 of `localbranching.py` - I'll include a screenshot of the code here. However, I know it is invoked several other times as well.

![image](https://github.com/user-attachments/assets/84a52043-fd83-447c-9ae7-1727de246106)

However I've seen other attributes, like `Model().copy_orig()` that I don't think are wrapped either. The repo specifies pySCIPopt 3.1.1, but I just can't see how the code is supposed to work, when the functions used seems like they haven't been defined. The repo have a open issue there on this as well, with no reaction from the authors, which is why I decided to check here, and see if there was something obvious I'd missed. 

They do ask you to compile a local version of Ecole from the repo (which I couldn't make work either), but I can't see how that would change the interaction between pySCIPopt and SCIP.

Thank you so much for your response!

**Reply by Joao-Dionisio** [reply, marked as answer]
Didn't know this existed, thanks @Engberg-Jacob! It seems that they are grabbing some extra functions from SCIP in [cos.cpp](https://github.com/pandat8/ML4LB/blob/001839ace3506c8410a30d1f4d3188a3cd95e2dd/ecole/libecole/src/scip/cons.cpp#L20), and possibly in other files as well. They are extending the functionality of PySCIPOpt this way, as PySCIPOpt doesn't have every SCIP function.

So yeah, one possibility for you is to wait until we add these functions (I'll talk with my colleagues). Another possibility is to install the local ecole version and ask the people from the repo for help, by creating an issue or sending an email.

**Reply by Engberg-Jacob** [reply]
Yeah, that's what I was thinking, just wanted to check for sure. Thank you!
